instrumentation/metric/registry.go

```diff
@@ -79,14 +89,46 @@ func (r *inMemoryRegistry) String() string {
 	return s
 }
 
-func (r *inMemoryRegistry) ExportAll() map[string]interface{} {
+func (r *inMemoryRegistry) ExportAll() map[string]exportedMetric {
 	r.mu.Lock()
 	defer r.mu.Unlock()
 
-	all := make(map[string]interface{})
+	all := make(map[string]exportedMetric)
 	for _, m := range r.mu.metrics {
 		all[m.Name()] = m.Export()
 	}
 
 	return all
 }
+
+func (r *inMemoryRegistry) ExportAndRotate() map[string]exportedMetric {
+	results := r.ExportAll()
+
+	for _, m := range r.mu.metrics {
+		m.Reset()
+	}
+
+	return results
+}
+
+func (r *inMemoryRegistry) report(logger log.BasicLogger) {
+	for _, value := range r.ExportAndRotate() {
+		logger.Metric("metric recorded", value.LogRow()...)
+	}
+}
+
+func (r *inMemoryRegistry) ReportEvery(ctx context.Context, interval time.Duration, logger log.BasicLogger) {
+	go func() {
+		periodicalTrigger := synchronization.NewPeriodicalTrigger(interval, func() {
+			r.report(logger)
+		})
+
+		periodicalTrigger.Start()
+
+		select {
+		case <-ctx.Done():
```

Comment:
this is ugly - either add the context to `NewPeriodicalTrigger` - which would make sense anyway, or do both polling and cancellation inline in the `select` statement.

I think after @jlevison's refactor of BlockSync, the new trigger might be irrelevant - if so, don't use it

Reply:
@jlevison promised to refactor `NewPeriodicalTrigger` to be context-aware, so I would prefer to stick with the trigger until it happens.

Reply:
ETA? if he's not doing it this week, just do it yourself :)

Reply:
This week.